api/desecapi/models.py

```diff
@@ -418,6 +444,83 @@ def save(self, *args, **kwargs):
+
+        records_canonical_presentation_format = []
+        for r in records_presentation_format:
+            try:
+                records_canonical_presentation_format.append(RR.canonical_presentation_format(r, rdtype))
+            except binascii.Error:
+                # e.g., odd-length string
+                _error(r, 'Cannot parse hexadecimal or base64 record contents')
+            except dns.exception.SyntaxError as e:
+                # e.g., A/127.0.0.999
+                if e.args[0] == 'string too long':
+                    excess = len(r) - 255
+                    _error(r, f'Record content {excess} chars too long.')
+                _error(r, 'Record syntax malformed')
+            except dns.name.NeedAbsoluteNameOrOrigin:
+                _error(r, 'Hostname must be fully qualified (i.e., end in a dot: "example.com.")')
+            except ValueError:
+                # e.g., string ("asdf") cannot be parsed into int on base 10
+                _error(r, 'Cannot parse record contents')
+            except Exception as e:
+                # TODO see what exceptions raise here for faulty input
+                raise e
+
+        records_canonical_presentation_format = set(records_canonical_presentation_format)
+        if len(records_canonical_presentation_format) < len(records_presentation_format):
+            # Duplicate record content
+            raise ValidationError(f'Duplicate records in RR set {self.type}/{self.name}.')
+
+        return records_canonical_presentation_format
+
+    def save_records(self, records):
+        """
+        Updates this RR set's resource records, discarding any old values.
+
+        Records are expected in presentation format and are converted to canonical
+        presentation format (e.g., 127.00.0.1 will be converted to 127.0.0.1).
+        Raises if a invalid set of records is provided.
+
+        This method triggers the following database queries:
+        - two SELECT queries for comparison of old with new records
+        - for each removed record, one DELETE query
+        - if at least record was added, a total of one INSERT query
+
+        Changes are saved to the database immediately.
+
+        :param records: list of records in representation format
+        """
+        records = self.clean_records(records)
+
+        # Remove RRs that we didn't see in the new list
+        removed_rrs = self.records.exclude(content__in=records)  # one SELECT
+        for rr in removed_rrs:
+            rr.delete()  # one DELETE query
+
+        # Figure out which entries in records have not changed
+        unchanged_rrs = self.records.filter(content__in=records)  # one SELECT
+        unchanged_content = [unchanged_rr.content for unchanged_rr in unchanged_rrs]
+        added_content = filter(lambda c: c not in unchanged_content, records)
+
+        rrs = [RR(rrset=self, content=content) for content in added_content]
+        RR.objects.bulk_create(rrs)  # One INSERT
```

Comment:
Should we return anything reasonable here, e.g. number of deleted and added records? Both are available in a straightforward manner from `QuerySet.delete()` and `QuerySet.bulk_create()`.

Reply:
This method is currently only used in (effectively) `RRsetSerializer.update` and `RRsetSerializer.create`, both return the `RRset` in question. I don't see how these methods could use those values.